src/ifcopenshell-python/ifcopenshell/validate.py

```diff
@@ -365,7 +412,61 @@ def validate(f, logger, express_rules=False):
 
         if not has_invalid_value:
             for i, (attr, val, is_derived) in enumerate(zip(attrs, values, entity.derived())):
-
+                try:
+                    type_of_attribute = attr.type_of_attribute()
+                    declared_type = type_of_attribute.declared_type()
+                    type_name = declared_type.name()
+
+                    if type_name == "IfcDateTime" and val:
```

Comment:
I would implement validation of https://ifc43-docs.standards.buildingsmart.org/IFC/RELEASE/IFC4x3/HTML/lexical/IfcDuration.htm in this PR as well. And really make sure that the code is as general and reusable as possible. So that validators are stored for example in a global mapping like:

```
type_validators = {
  'IfcDateTime: [parse_datetime, validate_datetime],
  'IfcDuration: [ ... ]
  ...
}
```